### PR TITLE
fix local tox for pyright/verifytypes to match ci run

### PR DESCRIPTION
**Pull request by kristapratico**
pyright/verifytypes do not work well with editable installs or mixing editable installs with sdist/whl installs. For example, if a library is installed via whl, but it's dependency (azure-core) is installed by editable install, the type checker won't be able to find/resolve azure-core. From what I can tell, when we run tox in CI, we always install all packages from whl/sdist and so both checks are happy. However, when we run tox locally, it looks like we do an editable install for azure-core from the dev_requirements.txt for most libraries.
@scbedd is my understanding of this correct? 

CI:
verifytypes installed: aiohttp==3.8.3,aiosignal==1.3.1,astroid==2.6.6,async-timeout==4.0.2,asynctest==0.13.0,attrs==22.2.0,**azure-core @ file:///mnt/vss/_work/1/s/sdk/textanalytics/azure-ai-textanalytics/.tmp_whl_dir/azure_core-1.26.3-py3-none-any.whl**

local:
verifytypes installed: aiohttp==3.8.3,aiosignal==1.3.1,astroid==2.6.6,async-timeout==4.0.2,atomicwrites==1.4.1,attrs==22.2.0,**azure-core @ file:///C:/Users/krpratic/azure-sdk-for-python/sdk/core/azure-core**

The difference above breaks running verifytypes locally (which @pvaneck pointed out yesterday). ~This PR changes the envs for pyright/verifytypes to not install the dev_requirements.txt. Eng sys tools still get installed separately and I don't expect the other deps found in dev_requirements.txt to be necessary for these checks since they are just doing static analysis. With this change, local / CI runs for pyright/verifytypes match.~

### CODE REVIEW

**Review by pvaneck — APPROVED** [review]
Thanks, works for me locally now 😃. I take it now if core has typing changings, these runs will need to wait for the next core release in order for core-related typing updates to be reflected?

**Comment by kristapratico** [comment]
> Thanks, works for me locally now 😃. I take it now if core has typing changings, these runs will need to wait for the next core release in order for core-related typing updates to be reflected?

You make a good point. While this PR enables local tox runs, it makes the CI runs use the published azure-core version too. Currently in CI, it builds the whl from the latest in main and installs it: https://dev.azure.com/azure-sdk/public/_build/results?buildId=2162544&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=23d29da4-00cb-5783-b17c-96bda25d9c52&l=31 which is _better_ IMO:

> azure-core @ file:///mnt/vss/_work/1/s/sdk/cognitivelanguage/azure-ai-language-conversations/.tmp_whl_dir/azure_core-1.26.4-py3-none-any.whl

I think it would be ideal if the local tox run built the whl in the tmp dir too and installed it, but I'll need to look at the code more to understand why it's different. I think I'm hesitant to make a change that worsens the CI check just to enable the local check (aka thanks for being patient with me, Paul 😆).

**Comment by pvaneck** [comment]
Ah okay. Good to know the CI still uses the latest core code. Not sure who else runs verifytypes locally, but this works for me since it still catches typing issues that stem from the service code (and actually runs, which is a plus 😄).

**Comment by kristapratico** [comment]
Yeah, sorry if that wasn't clear before -- CI is the source of truth right now. I don't want to change that, just want to get the local tox run to work/match it.

File: eng/tox/create_whl_for_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import fnmatch
+import tempfile
+from packaging.version import parse
+from ci_tools.parsing import ParsedSetup
+from ci_tools.build import create_package
+from ci_tools.variables import in_ci
+logging.getLogger().setLevel(logging.INFO)
+
+
+def install_dev_reqs(file, pkg_root, temp_dir):
```
Comment (scbedd) [inline eng/tox/create_whl_for_dependencies.py:24]: How close is this code to original? Worth the trouble to move under `ci_tools.functions` or a like namespace?

That way we can duplicate this code exactly once, then share in both locations.

Comment (scbedd) [reply eng/tox/create_whl_for_dependencies.py:24]: Yeah and when we move the common code, please type it + code comment description. No reason to do `param` strings if you don't wish, but the description + type inline help a ton when using this function later.

File: eng/tox/tox.ini [hunk 2]
```diff
@@ -156,6 +158,8 @@ commands =
       -p {toxinidir} \
       -w {envtmpdir} \
       --package-type sdist
+    {envbindir}/python {toxinidir}/../../../eng/tox/create_whl_for_dependencies.py \
```
Comment (scbedd) [inline eng/tox/tox.ini:161]: I read through `create_package_and_install`. There is a part where it checks for the target already being installed, but I have changed my mind about ordering. I think you're g2g to leave this where it is.

File: eng/tox/create_whl_for_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
```
Comment (scbedd) [inline eng/tox/create_whl_for_dependencies.py:1]: `create_whl_for_dependencies` also installs.

- `install_dependencies.py`?
- `install_dependency_whls.py`?

I won't block the PR based on this name though 👍

**Review by scbedd — APPROVED** [review]
Please move the common code to under `functions.py` within `azure-sdk-tools`. If you can update `tox_harness` to call the same, that would be awesome, but if not I'm totally cool with making it accessible for re-use later.

Other than those changes, am fine with this PR.

File: eng/tox/create_whl_for_dependencies.py (hunk 1 above)
Comment (kristapratico) [reply eng/tox/create_whl_for_dependencies.py:24]: The main differences with this code are 1) we don't overwrite the user's dev_requirements.txt when running locally and 2) we build _and install_ the whls. 

So we could update the original functions to only do those two things when not running in CI, move them to ci_tools.functions, and then share them across this script and the tox harness. Would you like me to do that?

File: eng/tox/create_whl_for_dependencies.py (hunk 3 above)
Comment (kristapratico) [reply eng/tox/create_whl_for_dependencies.py]: Renamed to create_dependencies_and_install.py to kind of mirror create_package_and_install.py

**Review by scbedd — APPROVED** [review]
Thank you for the refactor work @kristapratico ! Every little bit helps a ton! 

This looks great! :shipit:

**Pull request merged**